pkg/aws/ec2/api/cleanup/node_cleanup.go

```diff
@@ -0,0 +1,91 @@
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cleanup
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api"
+	ec2API "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api"
+	enicleanup "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api/cleanup/eni"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/go-logr/logr"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// NodeTerminationCleanerto handle resource cleanup at node termination
+type NodeTerminationCleaner struct {
+	NodeName   string
+	EC2Wrapper ec2API.EC2Wrapper
+	Log        logr.Logger
+}
+
+func (n *NodeTerminationCleaner) GetENITagFilters() []*ec2.Filter {
+	nodeFilter := enicleanup.CommonNetworkInterfaceFilters
+	nodeFilter = append(nodeFilter, []*ec2.Filter{
+		{
+			Name:   aws.String("tag:" + config.NetworkInterfaceNodenameKey),
+			Values: []*string{aws.String(n.NodeName)},
+		},
+	}...)
+	return nodeFilter
+}
+
+// DeleteLeakedResources deletes all available network interfaces on the node at node termination
+func (e *NodeTerminationCleaner) DeleteLeakedResources() error {
+	var errors []error
+	var nextToken *string
+
+	api.LeakedENINodeCleanupCnt.Set(0)
+	for {
+		describeNetworkInterfaceIp := &ec2.DescribeNetworkInterfacesInput{
+			Filters: e.GetENITagFilters(),
+			// TODO: set this to be configurable
+			MaxResults: aws.Int64(1000),
+			NextToken:  nextToken,
+		}
+		describeNetworkInterfaceOp, err := e.EC2Wrapper.DescribeNetworkInterfaces(describeNetworkInterfaceIp)
+		if err != nil {
+			return fmt.Errorf("failed to describe network interfaces, cleanup will be retried in next cycle: %v", err)
+		}
+
+		for _, networkInterface := range describeNetworkInterfaceOp.NetworkInterfaces {
+			_, err := e.EC2Wrapper.DeleteNetworkInterface(&ec2.DeleteNetworkInterfaceInput{
+				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+			})
+			if err != nil {
+				// append err and continue, we will retry deletion in the next period/reconcile
+				errors = append(errors, fmt.Errorf("failed to delete leaked network interface %v:%v", *networkInterface.NetworkInterfaceId, err))
+				e.Log.Error(err, "failed to delete the leaked network interface",
+					"id", *networkInterface.NetworkInterfaceId)
+				// increment leaked ENIs count
+				api.LeakedENINodeCleanupCnt.Inc()
+				continue
+			}
+			e.Log.Info("deleted leaked ENI successfully", "eni id", networkInterface.NetworkInterfaceId)
+		}
+
+		nextToken = describeNetworkInterfaceOp.NextToken
+		if nextToken == nil {
+			break
+		}
+
+	}
+	if len(errors) > 0 {
+		return fmt.Errorf("failed to delete atleast one leaked network interface, will be retried: %v", errors)
+	}
```

Comment:
Seems rudundant error log for me since you already log error for every time they happen in the for loop.